### PR TITLE
added option to verify server's cert fingerprint

### PR DESCRIPTION
**Pull request by fphammerle**
https://github.com/corecode/dma/issues/10

### CODE REVIEW

File: crypto.c [hunk 1]
```diff
@@ -78,6 +78,30 @@ init_cert_file(SSL_CTX *ctx, const char *path)
...
+		syslog(LOG_WARNING, "failed to load fingerprint of server's certicate: %s",
+			   ssl_errstr());
+		return 1;
+	}
+	if(fingerprint_len != SHA256_DIGEST_LENGTH) {
+		syslog(LOG_WARNING, "sha256 fingerprint has unexpected length of %d bytes",
+		       fingerprint_len);
+		return 1;
+	}
+	for(unsigned int i=0; i<SHA256_DIGEST_LENGTH; i++) {
```
Comment (corecode) [inline crypto.c]: memcmp?

File: crypto.c [hunk 2]
```diff
@@ -78,6 +78,30 @@ init_cert_file(SSL_CTX *ctx, const char *path)
...
 }
 
+unsigned int verify_server_fingerprint(const X509 *cert)
+{
+	unsigned char fingerprint[EVP_MAX_MD_SIZE] = {0};
+	unsigned int fingerprint_len = 0;
+	if(!X509_digest(cert, EVP_sha256(), fingerprint, &fingerprint_len)) {
+		syslog(LOG_WARNING, "failed to load fingerprint of server's certicate: %s",
+			   ssl_errstr());
+		return 1;
```
Comment (corecode) [inline crypto.c]: `return (1);`

File: dma.8 [hunk 3]
```diff
@@ -226,6 +226,8 @@ Uncomment if you want TLS/SSL secured transfer.
 Uncomment if you want to use STARTTLS.
 Only useful together with
 .Sq SECURETRANS .
+.It Ic FINGERPRINT Xo
+Optionally verify SHA256 fingerprint of smarthost's certificate.
```
Comment (corecode) [inline dma.8]: How about: "Pin the server certificate by specifying its SHA256 fingerprint.  Only makes sense if you use a smarthost."

File: dma.conf [hunk 4]
```diff
@@ -25,6 +25,9 @@
 # SECURETRANSFER)
 #STARTTLS
 
+# Optionally verify SHA256 fingerprint of smarthost's certificate.
```
Comment (corecode) [inline dma.conf]: (copy doc here)

File: crypto.c [hunk 5]
```diff
@@ -78,6 +78,30 @@ init_cert_file(SSL_CTX *ctx, const char *path)
 	return (0);
 }
 
+unsigned int verify_server_fingerprint(const X509 *cert)
```
Comment (corecode) [inline crypto.c]:
```
int
verify_server_fingerprint(...)
```

File: conf.c [hunk 6]
```diff
@@ -218,7 +218,22 @@ parse_conf(const char *config_path)
...
-		else if (strcmp(word, "OPPORTUNISTIC_TLS") == 0 && data == NULL)
+		else if (strcmp(word, "FINGERPRINT") == 0) {
+			if (strlen(data) != SHA256_DIGEST_LENGTH * 2) {
+				errlogx(EX_CONFIG, "invalid sha256 fingerprint length");
+			}
+			unsigned char *fingerprint = malloc(SHA256_DIGEST_LENGTH);
+			if (fingerprint == NULL) {
+				errlogx(EX_CONFIG, "fingerprint allocation failed");
+			}
+			for (unsigned int i=0; i<SHA256_DIGEST_LENGTH; i++) {
```
Comment (corecode) [inline conf.c]: spaces around `=` and `<`

**Comment by fphammerle** [comment]
Thanks for the quick response and review

I applied your suggestions

**Pull request merged**